# Classification with Naive Bayes

In [1]:
import pandas as pd

## The watermelon dataset

In the following we use the Watermelon dataset from Zhou's textbook, where each record represents a watermelon with its features (color, sound, texture, etc) and with a boolean value indicating if the watermelon is ripe or not.
Notice:
Input features have different types: *color, root, sound, texture, umbilicus, surface* are categorical (have a finite number of nominal values that can only enumerated and compared by equality operators), while *density, sugar* are numerical (continuous) and mathematical operators can be used to transform their values.

We can easily read this CSV dataset using pandas.

In [3]:
dataset = pd.read_csv('sample_data/watermelon-dataset.csv')
dataset

ID  color            root    sound          texture        umbilicus  \
0    1  green           curly  muffled            clear           hollow   
1    2   dark           curly     dull            clear           hollow   
2    3   dark           curly  muffled            clear           hollow   
3    4  green           curly     dull            clear           hollow   
4    5  light           curly  muffled            clear           hollow   
5    6  green  slightly curly  muffled            clear  slightly hollow   
6    7   dark  slightly curly  muffled  slightly blurry  slightly hollow   
7    8   dark  slightly curly  muffled            clear  slightly hollow   
8    9   dark  slightly curly     dull  slightly blurry  slightly hollow   
9   10  green        straight    crisp            clear             flat   
10  11  light        straight     dull           blurry             flat   
11  12  light           curly  muffled           blurry             flat   
12  13  green  slightly curly  muffled  slightly blurry           hollow   
13  14  light  slightly curly     dull  slightly blurry           hollow   
14  15   dark  slightly curly  muffled            clear  slightly hollow   
15  16  light           curly  muffled           blurry             flat   
16  17  green           curly     dull  slightly blurry  slightly hollow   

   surface  density  sugar   ripe  
0     hard    0.697  0.460   True  
1     hard    0.774  0.376   True  
2     hard    0.634  0.264   True  
3     hard    0.608  0.318   True  
4     hard    0.556  0.215   True  
5     soft    0.403  0.237   True  
6     soft    0.481  0.149   True  
7     hard    0.437  0.211   True  
8     hard    0.666  0.091  False  
9     soft    0.243  0.267  False  
10    hard    0.245  0.057  False  
11    soft    0.343  0.099  False  
12    hard    0.639  0.161  False  
13    hard    0.657  0.198  False  
14    soft    0.360  0.370  False  
15    hard    0.593  0.042  False  
16    hard    0.719  0.103  False

# Your work: what you have to do

- 1. Use `CategoricalNB` and `GaussianNB` from `scikit-learn` to build a Naive Bayes classifier on the watermelon dataset. For the estimated probabilities, consider the Laplace correction techniques.

- 2. Using the previously trained model, predict the ripe/unripe class of the test example proposed by the book:
*ID=T1, color=green, root=curly, sound=muffled, texture=clear, umbilicous=hollow, surface=hard, density=0.697, sugar=0.460. *

- 3. Giving the dataset size, consider now an hold-out technique in which you draw randomly a single test example and you train the model on the remaining examples. Train both naive Bayes and decision trees.

- 4. Build the confusion matrix and the ROC curves for naive Bayes and decision trees.

- 5. Implement statistical tests (e.g. `ttest_rel` in [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html)) to compare the naive Bayes models trained at 3. to Decision Trees on the same examples with an hold-out method. Have the observed differences in predictive performance statistical  significance?